crates/ruma-events/src/secret/request.rs

```diff
@@ -0,0 +1,381 @@
+                        return Err(de::Error::missing_field("request_id"));
+                    }
+                    Some(id) => id,
+                };
+
+                Ok(RequestToDeviceEventContent::new(
+                    request_action,
+                    requesting_device_id,
+                    request_id,
+                ))
+            }
+        }
+
+        const FIELDS: &[&str] = &["name", "action", "requesting_device_id", "request_id"];
+        deserializer.deserialize_struct(
+            "RequestToDeviceEventContent",
+            FIELDS,
+            RequestToDeviceEventContentVisitor,
+        )
+    }
+}
+
+/// Action for a *m.secret.request* event.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub enum RequestAction {
+    /// Request a secret by its name.
+    Request(SecretName),
+
+    /// Cancel a request for a secret.
+    RequestCancellation,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+impl Serialize for RequestAction {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Request(name) => {
+                let mut st = serializer.serialize_map(Some(2))?;
+                st.serialize_entry("name", name)?;
+                st.serialize_entry("action", "request")?;
+                st.end()
+            }
+            Self::RequestCancellation => {
+                let mut st = serializer.serialize_map(Some(1))?;
+                st.serialize_entry("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => serializer.serialize_str(custom),
+        }
+    }
+}
+
+/// The name of a secret.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, StringEnum)]
+pub enum SecretName {
+    /// Cross-signing master key (m.cross_signing.master).
+    #[ruma_enum(rename = "m.cross_signing.master")]
+    CrossSigningMasterKey,
+
+    /// Cross-signing user-signing key (m.cross_signing.user_signing).
+    #[ruma_enum(rename = "m.cross_signing.user_signing")]
+    CrossSigningUserSigningKey,
+
+    /// Cross-signing self-signing key (m.cross_signing.self_signing).
+    #[ruma_enum(rename = "m.cross_signing.self_signing")]
+    CrossSigningSelfSigningKey,
+
+    /// Recovery key (m.megolm_backup.v1).
+    #[ruma_enum(rename = "m.megolm_backup.v1")]
+    RecoveryKey,
+
+    #[doc(hidden)]
+    _Custom(String),
+}
+
+#[cfg(test)]
+mod test {
+    use super::{RequestAction, RequestToDeviceEventContent, SecretName};
+    use matches::assert_matches;
+    use serde_json::{from_value as from_json_value, json, to_value as to_json_value};
+
+    #[test]
+    fn secret_request_serialization() {
+        let content = RequestToDeviceEventContent::new(
+            RequestAction::Request(SecretName::_Custom("org.example.some.secret".into())),
```

Comment:
Please don't use private API in tests when it's not necessary. `SecretName` has a `From<&str>` / `From<String>` implementation that can be used here, right?

Reply:
I removed the uses of `_Custom` where I could, but I think there's no good way around it for the two new tests I added. I don't think it makes sense to have a `From<&str>` impl for `RequestAction`.